### DISCUSSION TITLE
Is there a way to set project's **Target** and **Runtime**?

### DISCUSSION THREAD

**Discussion by Joebeazelman** [alire-project/alire#1181] [Q&A]
Currently, you can add and remove dependencies of a project as well as indirectly configure the project using Alire and or the project's *toml* file.

For my HAL scripting, I would like to issue an Alire command to configure the **Target** and **Runtime** variables for a project file. 

Unfortunately, it unclear if these variables are supported. Can it be done and how?

**Comment by mosteo** [comment]
There's nothing in Alire about these variables for now. Perhaps you can set them using environment variables?

**Comment by Fabien-Chouteau** [comment]
Right now Alire doesn't know about targets and run-times and I think it's going to stay like that for the foreseeable future.

Can you explain a little bit more what you want to achieve?

**Reply by Joebeazelman** [reply]
For my HAL library builder, I would like to set these attributes in the generated project file according to the instructions in your blog. Currently, I am inserting them as text lines into the project file. This is less than ideal, since the user may want to rerun the code generator. To avoid duplicate attributes, my code needs to check if the attributes already have been inserted into the file from a previous code generation run. This involves searching for the attributes and replacing them. While my code works, I was hoping Alire would handle it for me either through a command or by modifying the **toml** file. Setting flags or setting environment variables makes little sense since the attributes are unlikely to change.

In my original code written in Ada, I was able to use one of the libraries for managing project files. I moved over to Python since it has more ready-made facilities and more flexible. I was able to rewrite it in an afternoon with no prior Python experience. If it proves useful, I may back-port it to Ada and release it as an Alire package.

All that's left is calling the **startup-gen** to generate the startup and linker scripts. I need to figure out where to find parameters needed by **startup-gen** a consistent manner in order to avoid having the user find and manually entering them into the project file. Some SVDs don't contain all the required information unfortunately. 

On another note, I thought about taking a shortcut by leveraging the community tested specification files already provided by platform.io. They practically have every board and MCU available as easy to parse text files. I looked at it briefly, but could find independent specifications for their supported MCUs, only as boards.

**Reply by Fabien-Chouteau** [reply]
Yeah I don't think Alire should get involved here. What you can do is put all the GPR variables that you generate in a separate abstract project file. So that you can always regenerate this file completely, no need to parse it. This is what we do for the crate config in Alire.

> All that's left is calling the startup-gen to generate the startup and linker scripts. I need to figure out where to find parameters needed by startup-gen a consistent manner in order to avoid having the user find and manually entering them into the project file. Some SVDs don't contain all the required information unfortunately.

I found something for that and actually I make a prototype of project generator with it. It's the [modm-devices](https://github.com/modm-io/modm-devices.git) project, and my prototype is here: https://github.com/Fabien-Chouteau/device_configs

**Reply by Joebeazelman** [reply]
Thanks a lot!  There's a ton of very valuable information here. The goal of my project is to provide a rapid onboarding process for Ada MCU development. I took a look at your repo and while I am not certain what direction your going, we seem to be going in the same direction. It would be a waste to duplicate each other's efforts going forward.

The first part of my generator successfully initializes a new library crate using Alire, adds HAL as a dependency, generates MCU specific code using Svd2Ada, and performs the initial build. Currently, SVD file path is passed in as a command line argument, but it would be even better to just pass in the MCU name and let the generator automagically find and download the SVD from an external repository. 

It appears you already completed the last part of my project. I couldn't find the generator script for the prototype projects. Perhaps we should combine forces and take my project to the finish line and release it. If you're interested, I can add you to my private so we can collaborate on it. 

As a warning, I'm not a Python programmer, I'm learning it as I'm developing the generator.

**Reply by Fabien-Chouteau** [reply]
Indeed we have a common goal.

I'am ok too look at your project, if/when I have time. If you can consider making it public it will be easier to collaborate. 
Also other people than me might want to join in.

**Reply by Joebeazelman** [reply]
Sure, but I don't want people using it just yet as it's a bit rough. I haven't tested it on Windows or Linux.

**Reply by Fabien-Chouteau** [reply]
Just write this in your README. To own anything to people trying beta stuff.

**Reply by Joebeazelman** [reply]
OK It's public and GPL3.0 with readme stating issues.
